packages/jsii-pacmak/lib/targets/golang/module.ts

```diff
@@ -1,64 +1,254 @@
     );
   }
 
+  public hasSubmodules(): boolean {
+    return Boolean(this.submodules.length);
+  }
+
+  public findType(fqn: string): ModuleType | undefined {
+    return findTypeInTree(this, fqn);
+  }
+
+  public emit(code: CodeMaker): void {
+    this.open(code);
+    code.line(`package ${this.packageName}`);
+    code.line();
+
+    code.line('import (');
+    code.indent();
+    code.line(`"${JSII_MODULE_NAME}"`);
+    if (this.dependencies.size > 0) {
+      this.dependencies.forEach((mod) => {
+        if (mod.packageName !== this.packageName) {
+          code.line(`"${mod.packageName}"`);
+        }
+      });
+    }
+    code.unindent();
+    code.line(')');
+    code.line();
+
+    this.emitTypes(code);
+    this.close(code);
+
+    this.emitSubmodules(code);
+  }
+
   public emitTypes(code: CodeMaker) {
     Object.values(this.types).forEach((type) => {
       type.emit(code);
     });
+  }
 
-    //     this.submodules.forEach(submodule => {
-    //       code.openFile(submodule.filename);
-    //       submodule.emitTypes(code);
-    //       code.closeFile(submodule.filename);
-    //     });
+  public emitSubmodules(code: CodeMaker) {
+    this.submodules.forEach((submodule) => {
+      submodule.emit(code);
+    });
   }
 }
 
-export class Submodule extends Module {
-  public constructor(assembly: Assembly, submodule: JsiiSubmodule) {
-    super(assembly, submodule);
+export class Submodule extends ModuleFile implements Module {
+  private readonly assembly: JsiiSubmodule;
+  public readonly packageName: string;
+  public readonly moduleName: string;
+  public readonly file: string;
+  public readonly filePath: string;
+  public readonly parent: Module;
+  public readonly root: RootModule;
+
+  public constructor(
+    root: RootModule,
+    parent: Module,
+    submodule: JsiiSubmodule,
+  ) {
+    const moduleName = submodule.name
+      .replace('@', '')
+      .replace(/[^a-z0-9]/gi, '');
+    const packageName = moduleName;
+    const filePath = `${parent.filePath}/${moduleName}`;
+    const file = `${filePath}.go`;
+    super(file);
+
+    this.assembly = submodule;
+    this.root = root;
+    this.parent = parent;
+    this.packageName = packageName;
+    this.moduleName = moduleName;
+    this.filePath = filePath;
+    this.file = file;
+  }
+
+  public get types(): ModuleTypes {
+    return buildModuleTypes(this, this.assembly.types);
+  }
+
+  public get submodules(): Submodule[] {
+    return buildSubmodules(this.root, this, this.assembly.submodules);
+  }
+
+  public hasSubmodules(): boolean {
+    return Boolean(this.submodules.length);
   }
 
-  public get filename(): string {
-    return 'submodulename';
+  public get dependencies(): Set<Module> {
+    return new Set(
+      this.types
+        .reduce(
+          (accum: Module[], t: ModuleType) => [...accum, ...t.dependencies],
+          [],
+        )
+        .filter((mod) => mod.packageName !== this.packageName),
```

Comment:
I [mentioned](https://github.com/aws/jsii/pull/1892#discussion_r470905365) this in the original PR, but I find this functional syntax kind of hard to read.

Reply:
Broke out `flatMap` as a util and separated extraction of the module name string for imports. Let me know if that helps. Other targets/modules are a mix of loops or map/filter/reduce. I don't really care so fine to change this to a loop if thats what we want to standardize on.